### PR TITLE
feat(controller): optional database migration

### PR DESCRIPTION
**Pull request by book987**
#4822

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: workflow/controller/config.go [hunk 1]
```diff
@@ -46,9 +46,13 @@ func (wfc *WorkflowController) updateConfig(v interface{}) error {
...
-		err = sqldb.NewMigrate(session, persistence.GetClusterName(), tableName).Exec(context.Background())
-		if err != nil {
-			return err
+		if !persistence.SkipDBMigration {
+			err = sqldb.NewMigrate(session, persistence.GetClusterName(), tableName).Exec(context.Background())
+			if err != nil {
+				return err
+			}
+		} else {
+			log.Warn("DB migration is disabled")
```
Comment (alexec) [inline workflow/controller/config.go:55]: I think this can be Info level

File: config/config.go [hunk 2]
```diff
@@ -145,11 +145,12 @@ type PersistConfig struct {
...
-	ClusterName    string            `json:"clusterName,omitempty"`
-	ConnectionPool *ConnectionPool   `json:"connectionPool,omitempty"`
-	PostgreSQL     *PostgreSQLConfig `json:"postgresql,omitempty"`
-	MySQL          *MySQLConfig      `json:"mysql,omitempty"`
+	ArchiveTTL      TTL               `json:"archiveTTL,omitempty"`
+	ClusterName     string            `json:"clusterName,omitempty"`
+	ConnectionPool  *ConnectionPool   `json:"connectionPool,omitempty"`
+	PostgreSQL      *PostgreSQLConfig `json:"postgresql,omitempty"`
+	MySQL           *MySQLConfig      `json:"mysql,omitempty"`
+	SkipDBMigration bool              `json:"skipDBMigration,omitempty"`
```
Comment (alexec) [inline config/config.go:153]: minor - can we call this `skipMigration`

**Review by alexec — COMMENTED** [review]
My thought on this is - how do users know what SQL needs to be applied? Could we modify migrate.go to with a print-only mode - i.e. print out the SQL statements, but not apply them?

**Comment by book987** [comment]
> My thought on this is - how do users know what SQL needs to be applied? Could we modify migrate.go to with a print-only mode - i.e. print out the SQL statements, but not apply them?

🤔 I think print out mode is
1. helpful when using Argo at first time, just get the output and apply somewhere
1. help little or useless when upgrading Argo, most of users don't want to drop and rebuild database i think, they need to find a way generating what they should apply. Even tracking changes from commits since last version is more efficent.
1. user should run another kubectl command to get what we print out, and filter SQL from all output by them self

And, like i mention at [#4822](https://github.com/argoproj/argo/issues/4822#issuecomment-759266974), i don't think hardcoded db migration in go source code is a good way. I think we can:
- use a seperated directory for db migration like Django do, new schema changes means new migration file, user can get schema chagnes by just seeking at migration directory
- or we can just provide a SQL file represent schema changes when releasing new version

But the ways i mention above is not friendly for user using Argo at first time, Maybe we needs further disign.

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
I think this is plenty good enough for MVP.

**Pull request merged**